cmd/dep/main.go

```diff
@@ -312,3 +334,66 @@ func (c *commentWriter) Write(p []byte) (int, error) {
 	}
 	return len(p), nil
 }
+
+type profile struct {
+	cpuProfile string
+
+	memProfile     string
+	memProfileRate int
+
+	mutexProfile         string
+	mutexProfileFraction int
+
+	// TODO(jbd): Add block profile and -trace.
+
+	f *os.File // file to write the profiling output to
+}
+
+func (p *profile) start() error {
+	switch {
+	case p.cpuProfile != "":
+		if err := p.createOutput(p.cpuProfile); err != nil {
+			return err
+		}
+		return pprof.StartCPUProfile(p.f)
+	case p.memProfile != "":
+		if p.memProfileRate > 0 {
+			runtime.MemProfileRate = p.memProfileRate
+		}
+		return p.createOutput(p.memProfile)
+	case p.mutexProfile != "":
+		if p.mutexProfileFraction > 0 {
+			runtime.SetMutexProfileFraction(p.mutexProfileFraction)
+		}
+		return p.createOutput(p.mutexProfile)
+	}
+	return nil
+}
+
+func (p *profile) finish() error {
+	if p.f == nil {
+		return nil
+	}
+	switch {
+	case p.cpuProfile != "":
+		pprof.StopCPUProfile()
+	case p.memProfile != "":
+		if err := pprof.WriteHeapProfile(p.f); err != nil {
+			return err
+		}
+	case p.mutexProfile != "":
+		if err := pprof.Lookup("mutex").WriteTo(p.f, 2); err != nil {
+			return err
+		}
+	}
+	return p.f.Close()
```

Comment:
Should this be `defer`ed before the switch so that Close() is still called if the cases error?